kauri/tree.go

```diff
@@ -0,0 +1,170 @@
+package kauri
+
+import (
+	"math"
+
+	"github.com/relab/hotstuff"
+)
+
+// MaxChild determines the branching factor of the tree.
+const MaxChild = 2
+
+// TreeConfiguration is an abstraction for a tree communication model.
+type TreeConfiguration interface {
+	InitializeWithPIDs(ids map[hotstuff.ID]int)
+	GetHeight() int
+	GetChildren() []hotstuff.ID
+	GetSubTreeNodes() []hotstuff.ID
+	GetParent() (hotstuff.ID, bool)
+}
+
+// FaultFreeTree implements a fault free tree configuration.
+type FaultFreeTree struct {
+	ID                  hotstuff.ID
+	ConfigurationLength int
+	height              int
+	idToPosMapping      map[hotstuff.ID]int
+	posToIDMapping      map[int]hotstuff.ID
+}
+
+// CreateTree Creates the tree configuration, currently only fault free tree configuration is supported.
+func CreateTree(configurationLength int, myID hotstuff.ID) TreeConfiguration {
+
+	if configurationLength <= 0 {
+		return nil
+	}
+	temp := configurationLength
+	temp = temp - 1 //root
+	height := 1
+	for i := 1; temp > 0; i++ {
+		temp = temp - int(math.Pow(MaxChild, float64(i)))
+		height++
+	}
+	return &FaultFreeTree{
+		ID:                  myID,
+		ConfigurationLength: configurationLength,
+		height:              height,
+	}
+}
+
+// InitializeWithPIDs uses the map to initialize the position of replicas.
+func (t *FaultFreeTree) InitializeWithPIDs(ids map[hotstuff.ID]int) {
+	t.idToPosMapping = ids
+	t.posToIDMapping = make(map[int]hotstuff.ID)
+	for ID, index := range ids {
+		t.posToIDMapping[index] = ID
+	}
+}
+
+// GetParent fetches the ID of the parent, if root, returns itself.
```

Comment:
In Go we typically don't use `Get` for getters... You can simply call this `Parent()`.

Also, the doc comment should maybe say something like:
```
Parent returns the tree node's parent ID and true.
Otherwise, if the tree node is the root, it returns its own ID and false.
```